README.md

```diff
@@ -2,11 +2,165 @@
+
+```ts
+@unmanaged
+export class ReferenceEntry {
+  rttid: u32; // The type Id
+  entryId: u32; // some kind of unique entry identifier
+  offset: usize; // and how big it is
+}
+```
+
+For every reference in an object tree, we can allocate some space in a buffer to describe it's shape. The `ReferenceEntry` object acts as a c-like pointer, pointing to each of the references in the buffer. The next step is to describe how they are linked using a `LinkEntry` class.
+
+```ts
+@unmanaged
+export class LinkEntry {
+  parentEntryId: u32; // The parent entry id
+  offset: usize; // Where it should be stored on the parent
+  childEntryId: u32; // The child entry id
+}
+```
+
+Since the links between parents and children are defined by way of this LinkEntry class, it makes it easy to describe circular references too. For instance, given a class like a `TreeNode` where each `TreeNode` can have a parent, and a collection of children, ASON will generate a `LinkEntry` that describes each relationship at `Serialization` time.
+
+
+```ts
+let a = new TreeNode();
+let b = new TreeNode();
+a.parent = b;
+b.children = [a];
+
+ASON.serialize(b);
+```
+
+This particular example happens to use another kind of `LinkEntry` called an `ArrayLinkEntry` because `children` is an array of objects, and linking objects to Arrays must be handled differently.
+
+However, the link entry table might conceptually look something like this as a JSON document:
+
+```ts
+{
+  "linkEntryTable": [
+    // the TreeNode
+    { parentEntryId: 0, childEntryId: 1, offset: offsetof<TreeNode>("parent") },
+    // the array
+    { parentEntryId: 1, childEntryId: 2, offset: offsetof<TreeNode>("children") },
+  ]
+}
+```
+
+Figuring out the shape of `ArrayLinkEntry` objects is left as an exercise to the reader.
+
+Lastly, we assert that the `entryId` stored at entry `0` is the primary entry a buffer. This of course assumes the generic type of the `Serializer` used is actually a reference.
+
+# Caveats
+
+When using the regular `ASON.serialize` method, values automatically will be boxed and unboxed for you.
+
+```ts
+assert(ASON.deserialize<i32>(ASON.serialize(42)) == 42);
+```
+
+However, when using the `Serializer` and `Deserializer` class directly, values must be boxed like this:
+
+```ts
+class Box<T> { constructor(public value: T) {} }
+
+let ser = new Serializer<f32>(); // Compile time error!
+
+// instead do this
+let ser = new ASON.Serializer<Box<f32>>();
+let des = new ASON.Deserializer<Box<f32>>();
+assert(des.deserialize<Box<f32>>(ser.serialize(new Box<f32>(42))).value == <f32>42);
+```
+
+ASON serialization optimizes for large object trees at the cost of making simple serialization slightly more expensive.
```

Comment:
This section is more important and should be elevated to the top to describe *why* using ASON requires reference serialization.